README.md
# flexi-task

## Usage

To launch run:

```bash
lein cljsbuild once
lein run

open http://localhost:3000
```

To run tests run:

```bash
lein test
```

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-08-06
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-08-06
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/flexi-task/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/flexi-task/compare/0.1.0...0.1.1


test/flexi_task/web_test.clj
(ns flexi-task.web-test
  (:require [clojure.test :refer :all]
            [flexi-task.web :refer :all]
            [ring.mock.request :as mock]))


(deftest scramble-handler-test
  (testing "Scramble handler without expected errors"
    (is (= (scramble-handler (-> (mock/request :post "/scramble")
                                 (mock/json-body {:str1 "bar", :str2 "baz"})))
           {:status  200
            :body    {:errors [] :answer false}}))
    (is (= (scramble-handler (-> (mock/request :post "/scramble")
                                 (mock/json-body {:str1 " bar ", :str2 "ba "})))
           {:status  200
            :body    {:errors [] :answer true}})))
  (is (= (scramble-handler (-> (mock/request :post "/scramble")
                               (mock/json-body {:str1 "", :str2 ""})))
         {:status  200
          :body    {:errors [] :answer true}}))

  (testing "Scramble handler with expected errors"
    (is (= (scramble-handler (-> (mock/request :post "/scramble")
                                 (mock/json-body {})))
           {:status  400
            :body    {:errors ["No value was found" "No value was found"] :answer false}}))
    (is (= (scramble-handler (-> (mock/request :post "/scramble")
                                 (mock/json-body {:str1 " bar99 ", :str2 "ba "})))
           {:status  400
            :body    {:errors ["Value can be only lower-case letters"] :answer false}}))))


test/flexi_task/core_test.clj
(ns flexi-task.core-test
  (:require [clojure.test :refer :all]
            [flexi-task.core :refer :all]))


(deftest scramble?-test
  (testing "scramble? works correctly"
    (are [x y] [= x y]
      true (scramble? "" "")
      true (scramble? "a" "")
      false (scramble? "" "a")
      false (scramble? "" "ax")
      true (scramble? "a" "a")
      true (scramble? "aa" "a")
      true (scramble? "aa" "aa")
      true (scramble? "ca" "ac")
      false (scramble? "a" "aa")
      true (scramble? "aba" "aa")
      true (scramble? "ababb" "b")
      true (scramble? "ababb" "aba")
      true (scramble? "fdsa" "asdf")
      false (scramble? "fdsa" "asdfu")
      true (scramble? "строка" "строка")
      true (scramble? "рссокккат" "строка")
      false (scramble? "рссоккка" "строка")
      true (scramble? "rekqodlw" "world")
      true (scramble? "cedewaraaossoqqyt" "codewars")
      false (scramble? "katas" "steak"))))

(deftest sanitize-test
  (testing "sanitize works correctly"
    (are [x y] [= x y]
      nil (sanitize nil)
      ["d"] (sanitize ["d"])
      "" (sanitize "")
      "as" (sanitize "as")
      "as" (sanitize "as    ")
      "as" (sanitize "    as")
      "as" (sanitize "    as ")
      "as sd" (sanitize "as sd")
      "as sd" (sanitize "  as sd   "))))

(deftest validate-test
  (testing "validate works correctly"
    (are [x y] [= x y]
      "No value was found" (validate nil)
      "Value is not a string" (validate ["abc"])
      "Value can be only lower-case letters" (validate "abc1")
      "Value can be only lower-case letters" (validate "ASDF")
      "Value can be only lower-case letters" (validate "asdf qwerty")
      nil (validate "")
      nil (validate "asdf")
      nil (validate "строка")
      "Value can be only lower-case letters" (validate "Cтрока"))))

test/flexi_task/views_test.clj
(ns flexi-task.views-test
  (:require [clojure.test :refer :all]
            [flexi-task.views :refer :all]))


(deftest index-page-contents-test
  (testing "index page should contain proper contents (just for sanity)"
    (is (.contains (index-page) "<title>FlexiTask</title>"))
    (is (.contains (index-page) "<h1>Let's play scramble!</h1>"))))

(deftest index-inputs-test
  (testing "index page should contain proper inputs"
    (is (.contains (index-page) "<input id=\"word1\""))
    (is (.contains (index-page) "<input id=\"word2\""))
    (is (.contains (index-page) "<input id=\"answer\""))
    (is (.contains (index-page) "<input id=\"send-btn\" onclick=\"flexi_task.core.check_scramble()\""))))


project.clj
(defproject flexi-task "Simple scramble Web App"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [compojure "1.6.2"]
                 [hiccup "1.0.5"]
                 [ring/ring-core "1.9.4"]
                 [ring/ring-jetty-adapter "1.9.4"]
                 [ring/ring-json "0.5.1"]
                 ; cljs deps
                 [org.clojure/clojurescript "1.10.866"]
                 [cljs-ajax "0.7.5"]
                 ; test deps
                 [ring/ring-mock "0.4.0"]]
  :repl-options {:init-ns flexi-task.core}
  :plugins [[lein-cljfmt "0.8.0"]
            [lein-cljsbuild "1.1.8"]]
  :cljsbuild {
    :builds [{:source-paths ["frontend"]
              :compiler {:output-to "resources/public/js/main.js"
                         :optimizations :whitespace
                         :pretty-print true}}]}
  :main flexi-task.web)


src/flexi_task/core.clj
(ns flexi-task.core
  (:require [clojure.string :as str]))


(defn sanitize
  "Sanitize input string"
  [x]
  (if (string? x)
    (str/trim x)
    x))

(defn validate
  "Validate input string"
  [x]
  (cond
    (nil? x) "No value was found"
    (not (string? x)) "Value is not a string"
    (not (re-matches #"\p{Ll}*" x)) "Value can be only lower-case letters"
    :else nil))

(defn scramble?
  "Checks if portion of A string characters can be rearranged to match B string.
  Performance notes:
  Has O(3N) complexity. Can be optimized for short-circuit cases and for O(2N) as well if needed."
  [a b]
  (let [b-freqs (atom (frequencies b))]
    (doseq [k     a
            :let  [v (get @b-freqs k 0)]
            :when (> v 0)]
      (swap! b-freqs update k dec))
    (empty? (filter (comp pos? val) @b-freqs))))


src/flexi_task/web.clj
(ns flexi-task.web
  (:use compojure.core
        flexi-task.core
        flexi-task.views)
  (:require [ring.adapter.jetty :as jetty]
            [ring.middleware.json :refer [wrap-json-response json-body-request]]
            [compojure.route :as route]))


(defn scramble-handler
  "Handler for scramble action. Handles input and checks for a scramble match"
  [req]
  (let [body (get (json-body-request req {}) :body)
        inputs (map (comp sanitize (partial get body)) ["str1" "str2"])
        validation-errors (filter some? (map validate inputs))]
    (if (empty? validation-errors)
      {:status 200 :body {:errors [] :answer (apply scramble? inputs)}}
      {:status 400 :body {:errors validation-errors :answer false}})))

(defroutes app-routes
  "Application Web and REST API routes definition"
  (GET "/" [] (index-page))
  (POST "/scramble" req (scramble-handler req))
  (route/resources "/")
  (route/not-found "Sorry, page was not found"))

(defn -main []
  (jetty/run-jetty (wrap-json-response app-routes) {:port 3000}))
